tests/host_callback_to_tf_test.py

```diff
@@ -0,0 +1,256 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""An example of using host_callback.call to invoke on the host functions
+written in Tensorflow. The interesting aspect here is how we can differentiate
+through the outside computation, using tf.GradientTape on the host.
+
+This is separate from host_callback_test because it needs a TF dependency.
+"""
+from typing import Callable
+import unittest
+
+from absl.testing import absltest
+from absl.testing import parameterized
+
+from jax import api
+from jax.config import config
+from jax import numpy as jnp
+from jax import test_util as jtu
+from jax.experimental import host_callback as hcb
+
+try:
+  import tensorflow as tf
+except ImportError:
+  tf = None
+
+config.parse_flags_with_absl()
+FLAGS = config.FLAGS
+
+
+def call_tf_no_ad(tf_fun: Callable, arg, *, result_shape):
+  """The simplest implementation of calling to TF, without AD support.
+
+  We must use hcb.call because the TF invocation must happen outside the
+  JAX staged computation."""
+
+  def tf_to_numpy(t):
+    return t.numpy() if isinstance(t, tf.Tensor) else t
+
+  return hcb.call(lambda arg: tf.nest.map_structure(tf_to_numpy,
+                                                    tf_fun(arg)),
+                  arg, result_shape=result_shape)
+
+
+def call_tf_simple_ad(tf_fun: Callable, arg, *, result_shape):
+  """Calls a TensorFlow function with simple support for reverse AD.
+
+  Works only for 1st order AD and only for arguments and results being a single
+  ndarray (no pytrees). Functions whose name starts with "tf_" are TensorFlow
+  functions and must be called outside the JAX computation.
+  """
+
+  @api.custom_vjp
+  def make_call(arg):
+    """We wrap it all in `make_call` so that we can attach custom VJP."""
+    return call_tf_no_ad(tf_fun, arg, result_shape=result_shape)
+
+  # Define the fwd and bwd custom_vjp functions
+  def make_call_vjp_fwd(arg):
+    # Return the primal argument as the residual. Use `make_call` for the
+    # primal computation to enable higher-order AD.
+    return make_call(arg), arg
+
+  def make_call_vjp_bwd(res, ct_res):
+    arg = res  # residual is the primal argument
+
+    def tf_vjp_fun(arg_and_ct_res):
+      """Invoke TF gradient; used with hcb.call."""
+      arg, ct_res = arg_and_ct_res
+      arg_var = tf.Variable(arg)
+      with tf.GradientTape(persistent=True) as tape:
+        res = tf_fun(arg_var)
+
+      dres_darg = tape.gradient(res, sources=arg_var,
+                                unconnected_gradients=tf.UnconnectedGradients.ZERO)
+      return dres_darg * ct_res
```

Comment:
You can let TensorFlow do this multiplication for you:
```suggestion
      dres_darg = tape.gradient(res, sources=arg_var,
                                output_gradients=ct_res,
                                unconnected_gradients=tf.UnconnectedGradients.ZERO)
      return dres_darg
```